File: packages/provide-uterm-go/vt/normalize.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

package vt

// NFC normalization, used when a combining character is drawn onto an
// existing cell (pyte calls unicodedata.normalize("NFC", ...)). This is a
// self-contained implementation over the generated canonical decomposition
// and primary composition tables plus algorithmic Hangul handling.

// Hangul syllable composition constants (Unicode chapter 3.12).
const (
	hangulSBase  = 0xAC00
	hangulLBase  = 0x1100
	hangulVBase  = 0x1161
	hangulTBase  = 0x11A7
	hangulLCount = 19
	hangulVCount = 21
	hangulTCount = 28
	hangulNCount = hangulVCount * hangulTCount
	hangulSCount = hangulLCount * hangulNCount
)

// canonicalDecomposeRune appends the full canonical decomposition of r.
func canonicalDecomposeRune(out []rune, r rune) []rune {
	if r >= hangulSBase && r < hangulSBase+hangulSCount {
		sIndex := r - hangulSBase
		out = append(out,
			hangulLBase+sIndex/hangulNCount,
			hangulVBase+(sIndex%hangulNCount)/hangulTCount)
		if t := sIndex % hangulTCount; t != 0 {
			out = append(out, hangulTBase+t)
		}
		return out
	}
	if d, ok := canonicalDecomp[r]; ok {
		out = canonicalDecomposeRune(out, d[0])
		if d[1] >= 0 {
			out = canonicalDecomposeRune(out, d[1])
		}
		return out
	}
	return append(out, r)
}

// canonicalOrder sorts sequences of non-starters by combining class using
// a stable bubble pass (the canonical ordering algorithm).
func canonicalOrder(rs []rune) {
	for i := 1; i < len(rs); i++ {
		cc := combiningClass(rs[i])
		if cc == 0 {
			continue
		}
		j := i
		for j > 0 {
			prev := combiningClass(rs[j-1])
			if prev == 0 || prev <= cc {
				break
			}
			rs[j-1], rs[j] = rs[j], rs[j-1]
			j--
		}
	}
}

// composePair returns the primary composite of a and b, if any.
func composePair(a, b rune) (rune, bool) {
	// Hangul L + V -> LV syllable.
	if a >= hangulLBase && a < hangulLBase+hangulLCount &&
		b >= hangulVBase && b < hangulVBase+hangulVCount {
		return hangulSBase + ((a-hangulLBase)*hangulVCount+(b-hangulVBase))*hangulTCount, true
	}
	// Hangul LV + T -> LVT syllable.
	if a >= hangulSBase && a < hangulSBase+hangulSCount && (a-hangulSBase)%hangulTCount == 0 &&
		b > hangulTBase && b < hangulTBase+hangulTCount {
		return a + (b - hangulTBase), true
	}
	c, ok := compositionPairs[compKey{a, b}]
	return c, ok
}

// composeRunes performs the canonical composition pass in place over a
// decomposed, canonically ordered sequence.
func composeRunes(rs []rune) []rune {
	if len(rs) == 0 {
		return rs
	}
	out := rs[:1]
	starter := -1
	lastCC := combiningClass(rs[0])
	if lastCC == 0 {
		starter = 0
	}
	for _, r := range rs[1:] {
		cc := combiningClass(r)
		// r is blocked from the starter when the preceding character is
		// a non-starter with a combining class >= cc.
		if starter >= 0 && (lastCC < cc || lastCC == 0) {
			if comp, ok := composePair(out[starter], r); ok {
				out[starter] = comp
				continue
			}
		}
		out = append(out, r)
		if cc == 0 {
			starter = len(out) - 1
		}
		lastCC = cc
	}
	return out
}

// nfcNormalize returns the NFC normalization of s, matching
// unicodedata.normalize("NFC", s).
func nfcNormalize(s string) string {
	decomposed := make([]rune, 0, len(s)+2)
	for _, r := range s {
		decomposed = canonicalDecomposeRune(decomposed, r)
	}
	canonicalOrder(decomposed)
	return string(composeRunes(decomposed))
}

File: packages/provide-uterm-csharp/src/Provide.Uterm/Vt/Normalize.cs
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

namespace Provide.Uterm.Vt;

internal static class Normalize
{
    private const int HangulSBase = 0xAC00;
    private const int HangulLBase = 0x1100;
    private const int HangulVBase = 0x1161;
    private const int HangulTBase = 0x11A7;
    private const int HangulLCount = 19;
    private const int HangulVCount = 21;
    private const int HangulTCount = 28;
    private const int HangulNCount = HangulVCount * HangulTCount;
    private const int HangulSCount = HangulLCount * HangulNCount;

    private static void CanonicalDecomposeRune(List<int> output, int r)
    {
        if (r >= HangulSBase && r < HangulSBase + HangulSCount)
        {
            var sIndex = r - HangulSBase;
            output.Add(HangulLBase + sIndex / HangulNCount);
            output.Add(HangulVBase + (sIndex % HangulNCount) / HangulTCount);
            var t = sIndex % HangulTCount;
            if (t != 0)
            {
                output.Add(HangulTBase + t);
            }

            return;
        }

        if (UnicodeNormTables.CanonicalDecomp.TryGetValue(r, out var d))
        {
            CanonicalDecomposeRune(output, d.Item1);
            if (d.Item2 >= 0)
            {
                CanonicalDecomposeRune(output, d.Item2);
            }

            return;
        }

        output.Add(r);
    }

    private static void CanonicalOrder(List<int> rs)
    {
        for (var i = 1; i < rs.Count; i++)
        {
            var cc = Wcwidth.CombiningClass(rs[i]);
            if (cc == 0)
            {
                continue;
            }

            var j = i;
            while (j > 0)
            {
                var prev = Wcwidth.CombiningClass(rs[j - 1]);
                if (prev == 0 || prev <= cc)
                {
                    break;
                }

                (rs[j - 1], rs[j]) = (rs[j], rs[j - 1]);
                j--;
            }
        }
    }

    private static bool ComposePair(int a, int b, out int comp)
    {
        if (a >= HangulLBase && a < HangulLBase + HangulLCount &&
            b >= HangulVBase && b < HangulVBase + HangulVCount)
        {
            comp = HangulSBase + ((a - HangulLBase) * HangulVCount + (b - HangulVBase)) * HangulTCount;
            return true;
        }

        if (a >= HangulSBase && a < HangulSBase + HangulSCount && (a - HangulSBase) % HangulTCount == 0 &&
            b > HangulTBase && b < HangulTBase + HangulTCount)
        {
            comp = a + (b - HangulTBase);
            return true;
        }

        return UnicodeNormTables.CompositionPairs.TryGetValue((a, b), out comp!);
    }

    private static List<int> ComposeRunes(List<int> rs)
    {
        if (rs.Count == 0)
        {
            return rs;
        }

        var output = new List<int> { rs[0] };
        var starter = -1;
        var lastCC = Wcwidth.CombiningClass(rs[0]);
        if (lastCC == 0)
        {
            starter = 0;
        }

        for (var idx = 1; idx < rs.Count; idx++)
        {
            var r = rs[idx];
            var cc = Wcwidth.CombiningClass(r);
            if (starter >= 0 && (lastCC < cc || lastCC == 0))
            {
                if (ComposePair(output[starter], r, out var composed))
                {
                    output[starter] = composed;
                    continue;
                }
            }

            output.Add(r);
            if (cc == 0)
            {
                starter = output.Count - 1;
            }

            lastCC = cc;
        }

        return output;
    }

    internal static string NfcNormalize(string s)
    {
        var decomposed = new List<int>(s.Length + 2);
        foreach (var rune in s.EnumerateRunes())
        {
            CanonicalDecomposeRune(decomposed, rune.Value);
        }

        CanonicalOrder(decomposed);
        var composed = ComposeRunes(decomposed);
        return string.Concat(composed.Select(r => char.ConvertFromUtf32(r)));
    }
}
